[Feature:Queue] Added multiple simultaneous queues to the office hours queue

elihschiff opened the pull request:
### Please check if the PR fulfills these requirements:

* [x] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)
* [x] Tests for the changes have been added/updated (if possible)
* [x] Documentation has been updated/added if relevant

### What is the current behavior?
<!-- List issue if it fixes/closes/implements one using the "Fixes #<number>" or "Closes #<number>" syntax -->
The main issue with the old queue was it only worked with one queue at a time

### What is the new behavior?

This now works with as many queues as you want at a time.
To test this code first make sure the office hour queue is enabled in the course settings. Go to the queue page and under the filter button add at least one queue, maybe 2-3. Then log in as a student and go to the same page to add yourself to the queue. Use the code you just typed as the name of the queue for the code. Now the mentor/ta can go and help the new student. You can also filter it to only show queues that you want to see under the filter button. 

![image](https://user-images.githubusercontent.com/18558130/72950873-9ab31500-3d5a-11ea-9a66-199e2ad46d15.png)
![image](https://user-images.githubusercontent.com/18558130/72950903-a7376d80-3d5a-11ea-9aa0-9a7dd22c34c6.png)
![image](https://user-images.githubusercontent.com/18558130/72950935-b9b1a700-3d5a-11ea-9225-559b1c49568c.png)
![image](https://user-images.githubusercontent.com/18558130/72950948-c33b0f00-3d5a-11ea-9020-9ccb9401ebe0.png)


### Other information?
<!-- Is this a breaking change? -->
<!-- How did you test -->
The migration script I wrote does work from my testing, but I don't know much sql so I don't know if there is a better way to do it.

Closes  #4667
This was mostly just a list of features I wanted to remember to add and not really a bug/issue. I added everything from there that I think needed to be added.

MasterOdin commented on migration/migrator/migrations/course/20200111145749_office_hours_queue_2.py:
```diff
@@ -0,0 +1,164 @@
+    #       2:Mentor/TA helped you
+    #       3:Mentor/TA removed you
+    #       4:Kicked out because queue emptied
+    #       5:You helped you
```
Let's break this apart into three columns instead of having it be contained in one column. It'll make it easier to potentially run future statistics over it, as well as just be a saner way of interacting with it.

MasterOdin commented on site/app/controllers/OfficeHoursQueueController.php:
```diff
@@ -32,31 +31,58 @@ public function showQueue() {
         }
+
+        // Print the entire match result
+        echo count($matches) != 1;
```
why is there an echo here?

elihschiff replied:
My bad that was from testing something. Good catch

MasterOdin commented on site/app/controllers/OfficeHoursQueueController.php:
```diff
@@ -67,146 +93,184 @@ public function addPerson() {
+        }
+        else {
+            $this->core->addSuccessMessage('Closed queue: "' . $_POST['queue_code'] . '"');
+        }
```
Let's make this a ternary statement with inline templating:
`$this->core->addSuccessMessage(($_POST['queue_state'] === 0 ? 'Opened' : 'Closed') . ' queue: "{$_POST['queue_code']}"');`

MasterOdin commented on site/app/controllers/OfficeHoursQueueController.php:
```diff
@@ -32,31 +31,58 @@ public function showQueue() {
+    }
+
+    /**
+    * @Route("/{_semester}/{_course}/office_hours_queue/open", methods={"POST"})
```
For many of these routes, let's add the `queue_code` as part of the route, as it's the direct resource that we're acting upon (which the route represents).

So, this one might be `/{_semester}/{_course}/office_hours_queue/{queue_code}/open`, etc.

The big change would be `/{_semester}/{_course}/office_hours_queue/new`, and I would just drop the `/new` bit and make POSTs to `/office_hour_queue` represent creation of new queues.

MasterOdin commented on site/app/controllers/OfficeHoursQueueController.php:
```diff
@@ -32,31 +31,58 @@ public function showQueue() {
+    * @return Response
+    */
+    public function openQueue() {
+        if (!isset($_POST['code']) || $_POST['code'] == "") {
```
Can replace the double check with just `if (empty($_POST['code']))`

MasterOdin commented on site/app/controllers/OfficeHoursQueueController.php:
```diff
@@ -67,146 +93,184 @@ public function addPerson() {
-        if (!isset($_POST['entry_id'])) {
-            $this->core->addErrorMessage("Missing entry ID");
+    public function removePerson() {
+        if (!isset($_POST['user_id'])) {
```
should this function also validate the code not being empty?

MasterOdin commented on site/app/controllers/OfficeHoursQueueController.php:
```diff
@@ -67,146 +93,184 @@ public function addPerson() {
-    public function finishHelpPerson() {
-        if (!isset($_POST['entry_id'])) {
-            $this->core->addErrorMessage("Missing entry ID");
+    public function startHelpPerson() {
```
should this function also validate the code not being empty?

MasterOdin commented on site/app/controllers/OfficeHoursQueueController.php:
```diff
@@ -67,146 +93,184 @@ public function addPerson() {
-    public function toggleQueue() {
-        if (!isset($_POST['queue_open'])) {
-            $this->core->addErrorMessage("Missing queue status");
+    public function finishHelpPerson() {
```
should this function also validate the code not being empty?

MasterOdin commented on site/app/controllers/OfficeHoursQueueController.php:
```diff
@@ -67,146 +93,184 @@ public function addPerson() {
     */
     public function emptyQueue() {
-        $this->core->getQueries()->emptyQueue($this->core->getUser()->getId());
+        if (!isset($_POST['queue_code']) && $_POST['queue_code'] != "") {
```
`empty($_POST['queue_code'])`

MasterOdin commented on site/app/controllers/OfficeHoursQueueController.php:
```diff
@@ -67,146 +93,184 @@ public function addPerson() {
-        if (!isset($_POST['entry_id'])) {
-            $this->core->addErrorMessage("Missing entry ID");
+    public function toggleQueue() {
+        if (!isset($_POST['queue_code']) && $_POST['queue_code'] != "") {
```
`empty($_POST['queue_code'])`

MasterOdin commented on site/app/controllers/OfficeHoursQueueController.php:
```diff
@@ -67,146 +93,184 @@ public function addPerson() {
         }
-        if (!$this->core->getUser()->accessGrading() && $this->core->getUser()->getId() != $this->core->getQueries()->getUserIdFromQueueSlot($_POST['entry_id'])) {
-            $this->core->addErrorMessage("Permission denied to remove that person");
+        if (!isset($_POST['queue_state']) && $_POST['queue_state'] != "") {
```
`empty($_POST['queue_state'])`

MasterOdin commented on site/app/controllers/OfficeHoursQueueController.php:
```diff
@@ -67,146 +93,184 @@ public function addPerson() {
-    public function generateNewCode() {
-        $this->core->getQueries()->genNewQueueCode();
+    public function deleteQueue() {
+        if (!isset($_POST['queue_code'])) {
```
`empty($_POST['queue_code'])`

MasterOdin commented on site/app/models/OfficeHoursQueueViewer.php:
```diff
@@ -0,0 +1,136 @@
+class OfficeHoursQueueViewer extends AbstractModel {
+
+
+    //If you want to see more details on the status codes see DatabaseQueries.php
```
I would like to see those status codes be enums within this model (or as a new model or something) and then all checks against specific numbers use those enum values, which makes it easier to follow the logic over time, and we are not relying on magic numbers anywhere.

elihschiff commented:
@MasterOdin I believe I fixed everything you have commented on so far


Also for the status codes, I broke them up into 2 separate columns in the DB. Because they are now split up I also switched them to strings. I think the strings are way more clear as instead of a magic number, you get a sense of what the code means. For example once somebody has been helped and is no longer in the queue, instead of being the value 2, it now is "helped" which I find much more clear.
Should I still make them into enums? I feel like there is really no point anymore because I would just be naming the variable names the exact same as the values.

Also thanks for the help reviewing my code
